codespell_lib/tests/test_dictionary.py

```diff
@@ -169,21 +171,78 @@ def test_error_checking_in_aspell(err, rep, err_aspell, rep_aspell, match):
         _check_err_rep(err, rep, (err_aspell, rep_aspell), 'dummy')
 
 
+# allow some duplicates, like "m-i-n-i-m-i-s-e", or "c-a-l-c-u-l-a-t-a-b-l-e"
+allowed_dups = {
+    ('dictionary.txt', 'dictionary_en-GB_to_en-US.txt'),
+    ('dictionary.txt', 'dictionary_rare.txt'),
+}
+
+
 @fname_params
+@pytest.mark.dependency(name='dictionary loop')
 def test_dictionary_looping(fname, in_aspell):
     """Test that all dictionary entries are valid."""
-    err_dict = dict()
+    this_err_dict = dict()
+    short_fname = op.basename(fname)
     with open(fname, 'rb') as fid:
         for line in fid:
             err, rep = line.decode('utf-8').split('->')
             err = err.lower()
-            assert err not in err_dict, 'error %r already exists' % err
+            assert err not in this_err_dict, \
+                'error %r already exists in %s' % (err, short_fname)
             rep = rep.rstrip('\n')
             reps = [r.strip() for r in rep.lower().split(',')]
             reps = [r for r in reps if len(r)]
-            err_dict[err] = reps
-    # check for corrections that are errors (but not self replacements)
-    for err in err_dict:
-        for r in err_dict[err]:
-            assert (r not in err_dict) or (r in err_dict[r]), \
-                ('error %s: correction %s is an error itself' % (err, r))
+            this_err_dict[err] = reps
+    # 1. check the dict against itself (diagonal)
+    for err in this_err_dict:
+        for r in this_err_dict[err]:
+            assert r not in this_err_dict, \
+                ('error %s: correction %s is an error itself in the same '
+                 'dictionary file %s' % (short_fname, err, r))
+    pair = (short_fname, short_fname)
+    assert pair not in global_pairs
+    global_pairs.add(pair)
+    for other_fname, other_err_dict in global_err_dicts.items():
+        # error duplication (eventually maybe we should just merge?)
```

Comment:
You could but you don't need to, the depends business is just a sanity check (it should always be true based on the logic here)

Reply:
Ah okay. But if we did that, we'd only get the error once wouldn't we? Aren't we currently doing some sort of weird increasing square process. On2?

Reply:
You have to check every dict against every other dict (including itself) so you're always stuck with `N * (N - 1) / 2`. If you think it's clearer to move it to the other one we can, six of one half a dozen of the other to me

Reply:
But given we're doing it within test_dictionary_looping which gets called for each dictionary, aren't we doing it factorially, or does your global_pairs stuff not avoid this?

Materially I think we should still be able to simplify it. If we can make a dict like so:
'1nd': ['dictionary', ['1st']]
'absolete': ['dictionary', ['absolute', 'obsolete']]
'tim': ['names' ['time']]

Then when adding them, if tim is already in the dict, that means the second dict duplicates the error, and we could even give the fact the first tim came from names. Then when checking corrections, if a correction matches an error we can give the dictionary the correction and the error are in.

I guess technically we could just have the dictionary name and then the corrections in one list, but I'd imagine the code would be clearer and less confusing if they were nested arrays or hashes or whatever.

Also if we lowercase on the way in we'd remove any issues with case. Given we already check if stuff is lower-case unique within one entry I think that would be okay.

Reply:
> But given we're doing it within test_dictionary_looping which gets called for each dictionary, aren't we doing it factorially, or does your global_pairs stuff not avoid this?

No it doesn't. On the first run, the global dict is empty so nothing is done other than comparing to itself (i.e., think of this as filling in the 0, 0 element in a matrix). Second run, the global dict has a single entry (the one from the first run), so first the 1, 1 entry is filled (second dict to itself), then 1, 0, and 0, 1. On the next, it fills the third row/col 2,0 2,1 2,2 1,2 0,2, etc. 

This seems simple enough to me but if it's to complicated to understand, you can just move all the checks to the `depends` test, it's explicit that it iterates over exactly every pair (actually `N * N` not `N * (N - 1) / 2` like I said previously, because it's directional) so if it's easier to "see" it there than above feel free to change it

Reply:
I should say no it doesn't do anything factorially. Just pairwise, as is necessary.

If you look at the `global_pairs` stuff it's just a sanity check -- first it checks to make sure that the given pair has not been compared, then compares them, and adds to the list of pairs that have been compared. So it ensures you only ever do `N * N` comparisons. And the `depends` test was actually added explicitly to try to make it clear that `N * N` pairwise comparisons had been done, and check that it's the case (though it always should be as long as nobody messes up the structure of the test)

Reply:
I think I got confused and thought your upper meant upper case. Not the best thing to try reviewing something complicated like this on a Night shift lunch break!

I'd assume the Python dict lookup will be slightly more efficient. More importantly now I'm all over the annotations, if we add the line number of each entry to my proposed format, we can output some errors that the annotator will catch and make it REALLY easy for people to see what errors they've made and where.